duneanalytics/spellbook
# Add cryptopunks to engine v2 + extend generic test framework

### Adding a Generic Test Framework

This code builds upon the work of @jeff-dude who introduced generic tests in the context of spellbook. This PR continues his work by adding a generic test that can be applied in many contexts: `tests/generic/compare_column_values_to_seed_values.sql`. This results in the following process:

#### Steps to add unit tests to a model
1. Create a CSV seed file with sample rows specific to the project you are integrating into spellbook.
E.g. `seeds/cryptopunks/cryptopunks_trades_buys_bids_seed.csv`
Every row in this file should contain expected values that you get from a trusted place such as a project's reporting. In this case, we used the [CryptoPunks project's website detail pages](https://cryptopunks.app/cryptopunks/details/5293). It's important here to **make sure that the column names in the seed file match those in the model**.

2. Add the test in the schema.yml file for the model
E.g. `models/cryptopunks/ethereum/cryptopunks_ethereum_schema.yml`
We added the following lines to test if `seller` values from seed and model match up:

```
- &seller
  name: seller
  description: "Seller wallet address"
  tests:
    - compare_column_values_to_seed_values:
        seed_file_location: ref('cryptopunks_trades_buys_bids_seed')
```


### Why is this PR WIP?

The transformation logic in [`cryptopunks_ethereum_events.sql`](./models/cryptopunks/ethereum/cryptopunks_ethereum_events.sql) needs a little more work review. I does not handle certain edge cases such as sales initiated by aggregators / vaults.


### General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test
* [x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`
* [x] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [x] if adding a new model, I edited the alter table macro to display new database object (table or view) in UI explorer
* [x] if adding a new materialized table, I edited the optimize table macro

Join logic:
* [x] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')
